services/web/server/tests/unit/with_dbs/fast/test_project_db.py

```diff
@@ -0,0 +1,651 @@
+
+    assert APP_PROJECT_DBAPI in client.app
+    db_api = client.app[APP_PROJECT_DBAPI]
+    assert db_api
+    # pylint:disable=protected-access
+    assert db_api._app == client.app
+    assert db_api._engine
+    return db_api
+
+
+async def test_setup_projects_db(client: TestClient):
+    _create_project_db(client)
+
+
+def test_project_db_engine_creation(postgres_db: sa.engine.Engine):
+    db_api = ProjectDBAPI.init_from_engine(postgres_db)
+    # pylint:disable=protected-access
+    assert db_api._app == {}
+    assert db_api._engine == postgres_db
+
+
+@pytest.fixture()
+async def db_api(client: TestClient, postgres_db: sa.engine.Engine) -> ProjectDBAPI:
+    db_api = _create_project_db(client)
+    yield db_api
+
+    # clean the projects
+    postgres_db.execute("DELETE FROM projects")
+
+
+def _assert_added_project(
+    exp_project: Dict[str, Any],
+    added_project: Dict[str, Any],
+    exp_overrides: Dict[str, Any],
+):
+    original_prj = deepcopy(exp_project)
+    added_prj = deepcopy(added_project)
+    # no user so the project owner has a pre-defined value
+    _DIFFERENT_KEYS = ["creationDate", "lastChangeDate"]
+
+    assert all(added_prj[k] != original_prj[k] for k in _DIFFERENT_KEYS)
+    assert to_datetime(added_prj["creationDate"]) > to_datetime(
+        exp_project["creationDate"]
+    )
+    assert to_datetime(added_prj["creationDate"]) <= to_datetime(
+        added_prj["lastChangeDate"]
+    )
+    original_prj.update(exp_overrides)
+    for k in _DIFFERENT_KEYS:
+        added_prj.pop(k)
+        original_prj.pop(k)
+    # the rest of the keys shall be the same as the original
+    assert added_prj == original_prj
+
+
+def _assert_project_db_row(
+    postgres_db: sa.engine.Engine, project: Dict[str, Any], **kwargs
+):
+    row: RowProxy = postgres_db.execute(
+        f"SELECT * FROM projects WHERE \"uuid\"='{project['uuid']}'"
+    ).fetchone()
+
+    expected_db_entries = {
+        "type": "STANDARD",
+        "uuid": project["uuid"],
+        "name": project["name"],
+        "description": project["description"],
+        "thumbnail": project["thumbnail"],
+        "prj_owner": None,
+        "workbench": project["workbench"],
+        "published": False,
+        "access_rights": {},
+        "dev": project["dev"],
+        "classifiers": project["classifiers"],
+        "ui": project["ui"],
+        "quality": project["quality"],
+        "creation_date": to_datetime(project["creationDate"]),
+        "last_change_date": to_datetime(project["lastChangeDate"]),
+    }
+    expected_db_entries.update(kwargs)
+    for k in expected_db_entries:
+        assert (
+            row[k] == expected_db_entries[k]
+        ), f"project column [{k}] does not correspond"
+    assert row["last_change_date"] >= row["creation_date"]
+
+
+@pytest.mark.parametrize(
+    "user_role",
+    [
+        # (UserRole.ANONYMOUS),
+        # (UserRole.GUEST),
+        (UserRole.USER),
+        # (UserRole.TESTER),
```

Comment:
still intend testing these roles?

Reply:
no they are here because the fixture needs them. removed.